src/uu/mv/src/mv.rs

```diff
@@ -726,3 +740,122 @@ fn is_empty_dir(path: &Path) -> bool {
         Err(_e) => false,
     }
 }
+
+/// Moves a directory from one location to another with progress tracking.
+///
+/// This is a modified version of `fs_extra::dir::move_dir_with_progress` tailored to fit our use case.
+/// This function assumes that `from` is a directory and `to` does not exist.
+///
+/// Parameters:
+/// - `from`: The source directory path.
+/// - `to`: The destination directory path.
+/// - `progress_bar`: An optional reference to a `ProgressBar` for tracking the movement progress.
+/// - `moved_files`: A mutable HashMap containing information about moved files.
+///
+/// Returns:
+/// - Result<u64>: The total number of bytes moved if successful.
+pub fn move_dir(
+    from: &Path,
+    to: &Path,
+    progress_bar: Option<&ProgressBar>,
+    moved_files: &mut HashMap<FileInformation, PathBuf>,
+) -> FsXResult<u64> {
+    if !from.exists() {
+        if let Some(msg) = from.to_str() {
+            let msg = format!("Path \"{}\" does not exist or you don't have access!", msg);
+            return Err(FsXError::new(FsXErrorKind::NotFound, &msg));
+        }
+        return Err(FsXError::new(
+            FsXErrorKind::NotFound,
+            "Path does not exist or you don't have access!",
+        ));
+    }
+
+    let dir_content = get_dir_content(from)?;
```

Comment:
`get_dir_content` does not return sufficient information. GNU mv copies/moves as many files/directories as it can, even if it first encounter some files where it has insufficient permissions:

```console
$ mv -v source/ /tmp
created directory '/tmp/source'
created directory '/tmp/source/d1'
mv: cannot access 'source/d1': Permission denied
created directory '/tmp/source/d3'
mv: cannot access 'source/d3': Permission denied
copied 'source/f2' -> '/tmp/source/f2'
[$? = 1]
$ ls source/  # source has NOT been deleted!
d1  d3  f2
$ ls /tmp/source/
d1  d3  f2
```

However, `get_dir_content` bails completely if there are *any* permission issues:

```console
$ cargo run -q --features mv mv -v source/ /tmp
mv: cannot move 'source/' to '/tmp/source': Permission denied
[$? = 1]
$ ls /tmp/source/
ls: cannot access '/tmp/source/': No such file or directory
[$? = 2]
```

And the final nail in the coffin is that `get_dir_content` insists to first read the entire filesystem structure, which uses up potentially vast amounts of memory, and gives ample opportunity for all kinds of TOCTOU races: https://docs.rs/fs_extra/latest/fs_extra/dir/struct.DirContent.html

Therefore, we should avoid using `get_dir_content` entirely. I'm not even sure it *can* be used correctly.

Reply:
Yes, you're right. Instead of using `get_dir_content`, we could follow the approach used in `uu_cp`, which involves walking through the directory, copying files, and creating folders as we go.